llvm/tools/sycl-post-link/ModuleSplitter.cpp

```diff
@@ -761,5 +761,146 @@ getDoubleGRFSplitter(ModuleDesc &&MD, bool EmitOnlyKernelsAsEntryPoints) {
     return std::make_unique<ModuleCopier>(std::move(MD), std::move(Groups));
 }
 
+namespace {
+  // Data structure, which represent a combination of all possible optional
+  // features used in a function.
+  //
+  // It has extra methods to be useable as a key in llvm::DenseMap.
+  struct UsedOptionalFeatures {
+    SmallVector<int, 4> Aspects;
+    // TODO: extend this further with reqd-sub-group-size, reqd-work-group-size,
+    // double-grf and other properties
+
+    UsedOptionalFeatures() = default;
+
+    UsedOptionalFeatures(const Function *F) {
+      if (const MDNode *MDN = F->getMetadata("sycl_used_aspects")) {
+        auto ExtractIntegerFromMDNodeOperand = [=](const MDNode *N,
+                                                   unsigned OpNo) -> auto {
+          Constant *C =
+              cast<ConstantAsMetadata>(N->getOperand(OpNo).get())->getValue();
+          return C->getUniqueInteger().getSExtValue();
+        };
+
+        for (size_t I = 0, E = MDN->getNumOperands(); I < E; ++I) {
+          Aspects.push_back(ExtractIntegerFromMDNodeOperand(MDN, I));
```

Comment:
Are duplicates possible here? If not, a comment would be appreciated (nit).

Reply:
Duplicates are not expected here. Added a comment in b39798bc075020de5c2621e0254a8dac46466f6d